### PR TITLE
ARTEMIS-1416 Queue is not autocreated if address already exists

### PR DESCRIPTION
**Pull request by gaohoward**
- Fix on core and amqp
- Add test to verify amqp's current large message behavior
- Add test to openwire also just to verify

### CODE REVIEW

**Comment by gaohoward** [comment]
looking...

**Comment by clebertsuconic** [comment]
#1625 seems related to this

**Comment by clebertsuconic** [comment]
Running these tests solved the AMQP tests I was looking at...

However, I now see this failure:

23:30:16,181 WARN  [org.apache.activemq.artemis.protocol.amqp.proton.handler.ProtonHandler] Incorrect Routing Type for queue, expecting: MULTICAST: ActiveMQAMQPInternalErrorException[errorType=INTERNAL_ERROR message=Incorrect Routing Type for queue, expecting: MULTICAST]
        at org.apache.activemq.artemis.protocol.amqp.proton.ProtonServerSenderContext.close(ProtonServerSenderContext.java:529) [artemis-amqp-protocol-2.4.0-SNAPSHOT.jar:]
        at org.apache.activemq.artemis.protocol.amqp.proton.AMQPConnectionContext.onRemoteClose(AMQPConnectionContext.java:474) [artemis-amqp-protocol-2.4.0-SNAPSHOT.jar:]
        at org.apache.activemq.artemis.protocol.amqp.proton.handler.Events.dispatch(Events.java:74) [artemis-amqp-protocol-2.4.0-SNAPSHOT.jar:]
        at org.apache.activemq.artemis.protocol.amqp.proton.handler.ProtonHandler.dispatch(ProtonHandler.java:450) [artemis-amqp-protocol-2.4.0-SNAPSHOT.jar:]
        at org.apache.activemq.artemis.protocol.amqp.proton.handler.ProtonHandler.flush(ProtonHandler.java:279) [artemis-amqp-protocol-2.4.0-SNAPSHOT.jar:]
        at org.apache.activemq.artemis.protocol.amqp.proton.handler.ProtonHandler.inputBuffer(ProtonHandler.java:243) [artemis-amqp-protocol-2.4.0-SNAPSHOT.jar:]
        at org.apache.activemq.artemis.protocol.amqp.proton.AMQPConnectionContext.inputBuffer(AMQPConnectionContext.java:150) [artemis-amqp-protocol-2.4.0-SNAPSHOT.jar:]
        at org.apache.activemq.artemis.protocol.amqp.broker.ActiveMQProtonRemotingConnection.bufferReceived(ActiveMQProtonRemotingConnection.java:141) [artemis-amqp-protocol-2.4.0-SNAPSHOT.jar:]
        at org.apache.activemq.artemis.core.remoting.server.impl.RemotingServiceImpl$DelegatingBufferHandler.bufferReceived(RemotingServiceImpl.java:642) [artemis-server-2.4.0-SNAPSHOT.jar:2.4.0-SNAPSHOT]
        at org.apache.activemq.artemis.core.remoting.impl.netty.ActiveMQChannelHandler.channelRead(ActiveMQChannelHandler.java:68) [artemis-core-client-2.4.0-SNAPSHOT.jar:2.4.0-SNAPSHOT]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-all-4.1.16.Final.jar:4.1.16.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-all-4.1.16.Final.jar:4.1.16.Final]
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) [netty-all-4.1.16.Final.jar:4.1.16.Final]
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1359) [netty-all-4.1.16.Final.jar:4.1.16.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-all-4.1.16.Final.jar:4.1.16.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-all-4.1.16.Final.jar:4.1.16.Final]
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:935) [netty-all-4.1.16.Final.jar:4.1.16.Final]
        at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:797) [netty-all-4.1.16.Final.jar:4.1.16.Final]
        at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:404) [netty-all-4.1.16.Final.jar:4.1.16.Final]
        at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:304) [netty-all-4.1.16.Final.jar:4.1.16.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858) [netty-all-4.1.16.Final.jar:4.1.16.Final]
        at java.lang.Thread.run(Thread.java:745) [rt.jar:1.8.0_121]




Which we need to address before merging this.


I have a dentist appoint the next day from now though.. so i will be back in a couple hours.

**Comment by clebertsuconic** [comment]
the failure at jenkins is accurate

**Comment by gaohoward** [comment]
turns out my PR indeed broke the other tests. I have a fix for that and I'll update the PR soon, after making sure there is no more regressions.

**Comment by clebertsuconic** [comment]
@gaohoward if you can just update the PR? This is exactly what I was testing.. so I would do the verifications.

On this case please keep your committs separate please? just add a new commit instead of ammend.

**Comment by gaohoward** [comment]
@clebertsuconic ok, I'll do it.

File: artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/broker/AMQPSessionCallback.java [hunk 1]
```diff
@@ -405,7 +407,10 @@ public void serverSend(final Transaction transaction,
       }
 
       //here check queue-autocreation
-      if (!bindingQuery(address, RoutingType.ANYCAST)) {
+      org.apache.qpid.proton.amqp.messaging.Target target = (org.apache.qpid.proton.amqp.messaging.Target) receiver.getRemoteTarget();
```
Comment (clebertsuconic) [inline artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/broker/AMQPSessionCallback.java]: ^^ this is not being used... I will amend without this as I merge

**Comment by clebertsuconic** [comment]
Please.. do not merge this Pull Request.. let me do some tests with it...

@gaohoward  nothing wrong with it.. it's actually great.. helped me a lot... just that what I'm doing is crossing a lot with this.. and I want to make sure it's all good.

**Comment by gaohoward** [comment]
@clebertsuconic thanks!

**Pull request closed without merging**